docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,322 @@
+
+Alice is a developer who wants to batch operations with ORAS in her Shell script. In order to automate some routine workflow in containers secure supply chain scenario, she wants the machine to get the image digest from the JSON output objects that are emitted by `oras push`, then use utility like [xargs](https://en.wikipedia.org/wiki/Xargs) or use environmental variables to enable an ORAS command to act on the output of another command and perform further steps. In this way, she can chain commands together. For example, she can use `oras attach` to attach an SBOM to the image using its image digest as a argument outputted from `oras push`.
+
+For example, push an artifact to a registry and generate the artifact reference in the standard output, then attach an SBOM to the artifact using the artifact reference (`$REGISTRY/$REPO@$DIGEST`) outputted from the first command, finally sign the attached SBOM with another tool against the SBOM file's reference (`$REGISTRY/$REPO@$DIGEST`) outputted from the last step.
+
+- Use xargs utility on Unix
+
+```bash
+oras push $REGISTRY/$REPO:$TAG hello.txt --format '{{.Ref}}' |\
+xargs -I _ oras attach --artifact-type sbom/example _ sbom.spdx --format '{{.Ref}}' |\
+xargs -I _ notation sign _ 
+```
+
+- Use environmental variables on Unix
+
+```bash
+REFERENCE_A=$(oras push $REGISTRY/$REPO:$TAG hello.txt --format '{{.Ref}}')
+REFERENCE_B=$(oras attach --artifact-type sbom/example $REFERENCE_A sbom.spdx --format '{{.Ref}}') 
+notation sign $REFERENCE_B
+```
+
+- Use [ConverFrom-Json](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json) on Windows PowerShell
+
+```powershell
+$REFERENCE_A=oras push $REGISTRY/$REPO:$TAG hello.txt --format json --no-tty | ConvertFrom-Json
+$REFERENCE_B=oras attach --artifact-type sbom/example $REFERENCE_A.Ref sbom.spdx --format json --no-tty | ConvertFrom-Json
+notation sign $REFERENCE_B.Ref
+```
+
+### CI/CD
+
+Bob is a DevOps engineer. He uses the ORAS GitHub Actions [Setup action](https://github.com/oras-project/setup-oras) to install ORAS in his CI/CD workflow. He wants to chain multiple ORAS commands in a Shell script to perform multiple operations.
+
+For example, pull multiple files (layers) from a repository and filter out the file path of its first layer in the standard output, then pass the pulled first layer to the second command `docker import` for further operation. 
+
+```yaml
+jobs:
+  example-job:
+    steps:
+      - uses: oras-project/setup-oras@v1
+      - run: |
+          PATH=`oras pull $REGISTRY/$REPO:$TAG --format '{{.first .Files.Path}}'`
+          docker import $PATH
+```
+
+## Proposal and desired user experience
+
+Enable users to use the `--format <type>` flag to format output into structured data (e.g. JSON) and use the `--format` with [Go template](https://pkg.go.dev/text/template) to manipulate the output fields. Users can still use `--format '{{toRawJson .}}'` to get raw JSON output.
+
+### oras pull 
```

Comment:
Could you include an example of pulling a folder?

Reply:
Added a Note at line184 to clarify this case